### ISSUE TITLE
[jit] training attribute broken in recursive script

### ISSUE THREAD

**Issue by tamas-the-sorcerer** [pytorch/pytorch#26884]
## 🐛 Bug

the model was scripted in eval mode and then saved with torch.jit.save. 
torch.jit.load loads the model in training mode. 
this is a problem, because in training mode dropouts are enabled, and also batchnorm needs to be adjusted. we expected torch.jit.load to load the model in the mode it was saved in. please fix this. 

## To Reproduce

**Steps to reproduce the behavior:**
```
tamas@super-duper-compute-machine:~/JUPYTER_LAB$ ipython
Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.6.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch                                                                                                                              

In [2]: torch.__version__                                                                                                                         
Out[2]: '1.2.0'

In [3]: class MyCell(torch.nn.Module): 
   ...:     def __init__(self): 
   ...:         super(MyCell, self).__init__() 
   ...:         self.linear = torch.nn.Linear(4, 4) 
   ...:  
   ...:     def forward(self, x, h): 
   ...:         new_h = torch.tanh(self.linear(x) + h) 
   ...:         return new_h, new_h 
   ...:  
   ...: model = MyCell() 
   ...: x, h = torch.rand(3, 4), torch.rand(3, 4) 
   ...: scripted_model = torch.jit.script(model, (x, h))                                                                                          
/home/tamas/anaconda3/envs/cuda/lib/python3.7/site-packages/torch/jit/__init__.py:1158: UserWarning: `optimize` is deprecated and has no effect. Use `with torch.jit.optimized_execution() instead
  warnings.warn("`optimize` is deprecated and has no effect. Use `with torch.jit.optimized_execution() instead")

In [4]: scripted_model.code                                                                                                                       
Out[4]: 'def forward(self,\n    x: Tensor,\n    h: Tensor) -> Tuple[Tensor, Tensor]:\n  _0 = self.linear\n  _1 = _0.weight\n  _2 = _0.bias\n  if torch.eq(torch.dim(x), 2):\n    _3 = torch.__isnot__(_2, None)\n  else:\n    _3 = False\n  if _3:\n    bias = ops.prim.unchecked_unwrap_optional(_2)\n    ret = torch.addmm(bias, x, torch.t(_1), beta=1, alpha=1)\n  else:\n    output = torch.matmul(x, torch.t(_1))\n    if torch.__isnot__(_2, None):\n      bias0 = ops.prim.unchecked_unwrap_optional(_2)\n      output0 = torch.add_(output, bias0, alpha=1)\n    else:\n      output0 = output\n    ret = output0\n  new_h = torch.tanh(torch.add(ret, h, alpha=1))\n  return (new_h, new_h)\n'

In [5]: scripted_model.training                                                                                                                   
Out[5]: True

In [6]: scripted_model.eval()                                                                                                                     
Out[6]: 
WeakScriptModuleProxy(
  (linear): WeakScriptModuleProxy()
)

In [7]: scripted_model.training                                                                                                                   
Out[7]: False

In [8]: torch.jit.save(scripted_model, 'temp.pt')                                                                                                 

In [9]: m = torch.jit.load('temp.pt')                                                                                                             

In [10]: m.training                                                                                                                               
Out[10]: True
```

cc @ezyang @gchanan @zou3519 @suo

**Comment by suo** [comment]
Thanks for the report, I can confirm this is indeed a bug. There are a few interacting problems:
1. Recursive script skips the training attributes (??)
2. This training attribute can potentially be added on-demand if requested during compilation (??)
3. Even if it's appropriately added, the call to `m.training` after load will return True because we've allowed the Python `training` attribute to shadow the `torch._C.ScriptModule` attribute.

@driazati can you take a look?

**Comment by eellison** [comment]
@driazati is on pto for the next two days, if this is high priority or you want to get it in for the release someone else needs to pick it up

**Comment by suo** [comment]
cc @zdevito, do you have time to look into this before the branch cut?

**Comment by driazati** [comment]
This looks like it's just an issue with accessing the `training` attribute from Python, if you run a script function that uses `training` it will be correctly set (which may end up being different from the value you see in Python) and things like dropout should work right. I'll have a PR to fix up the inconsistency soon.

**Comment by dzhulgakov** [comment]
It looks pretty bad and will cause a lot of confusion to users because .eval() + script() makes sense and is the thing they'd naturally do. Can we implement at least a workaround that calls eval inside script() wrapper if proper fix is hard. For quantized models it's bad because it fails loudly (#27205) but it's way worse for fp32 models because you'd get wrong outputs :(

**Comment by raghuramank100** [comment]
I second that, a work around before the release will be very helpful
